src/main/java/net/uaznia/lukanus/hudson/plugins/gitparameter/GitParameterDefinition.java

```diff
@@ -358,7 +394,50 @@ private boolean notMatchUseRepository(String gitUrl) {
         }
         return tagSet;
     }
+    
+
+    private Map<String, String> getAnnotatedTag(JobWrapper jobWrapper, GitSCM git, Map<String, String> paramList, EnvVars environment, RemoteConfig repository, URIish remoteURL) throws IOException, InterruptedException {
+    	boolean isRepoScm = RepoSCM.isRepoSCM(repository.getName());
+        FilePathWrapper workspace = getWorkspace(jobWrapper, isRepoScm);
 
+        GitClient gitClient = getGitClient(jobWrapper, workspace, git, environment);
+        initWorkspace(workspace, gitClient, remoteURL);
+        FetchCommand fetch = gitClient.fetch_().prune().from(remoteURL, repository.getFetchRefSpecs());
+        fetch.execute();
+        
+    	Map<String, String>  tagMap = new HashMap<String, String>();
+        try {
+            Set<String> tags = gitClient.getTagNames(tagFilter);
+            for (String tagName : tags) {
+        		tagName = tagName.replaceFirst(REFS_TAGS_PATTERN, "");
+        		String tagMessage = getTagMessage(gitClient, tagName);
+        		tagMessage = trimMessage(tagMessage);
+        		if(tagMessage.isEmpty()) {
+        			tagMessage = tagName;
+        		} else {
+        			tagMessage = tagName.concat(TAG_MESSAGE_SEPARATOR).concat(tagMessage);
+        		}
+        		tagMap.put(tagMessage, tagName);
+            }
+        } catch (GitException e) {
+            LOGGER.log(Level.WARNING, getCustomeJobName() + " " + Messages.GitParameterDefinition_getTag(), e);
+        }
+        
+        return tagMap;
+    }
+    private String getTagMessage(GitClient gitClient, String tagName) throws GitException, InterruptedException {
+    	return gitClient.getTagMessage(tagName).trim();
+    }
+    
```

Comment:
Both of these methods should be extract to RevisionInfoFactory
or create AnnotatedTagFactory to extract all logic for create message, 
a common method of RevisionInfoFactory and AnnotatedTagFactory (eg. trimMessage) should be extract to abstract class